File: 트랜스폼컨쿼/applicationProblem.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#define MAX_NODE 100
#define SWAP(a,b,t) {t=a;a=b;b=t;}
typedef struct node{
    int data;
}Node;
typedef struct heap{
    char name[31];
    int size;
    Node heap[MAX_NODE];
}Heap;
typedef struct heapSET{
    Heap* heaplist[MAX_NODE];
    int size;
}HeapSET;
Heap* initheap(char*);
void swap(Heap*,int,int);
void input(Heap*,int);
int delete(Heap*);
void bubbleSort(char**,int*);
int N;
int main(void){
    char* temp=(char*)malloc(sizeof(char)*31);
    char cmd;
    HeapSET* hset=(HeapSET*)malloc(sizeof(HeapSET));
    hset->size=0;
    for(int i=0;i<MAX_NODE;i++){
        hset->heaplist[i]=NULL;
    }
    scanf("%d",&N);
    char** names=(char**)malloc(sizeof(char*)*(N+1));
    int* expDate=(int*)malloc(sizeof(int)*N);
    for(int i=0;i<N;i++){
        names[i]=(char*)malloc(sizeof(char)*31);//물품이름은 10글자까지
    }
    for(int i=0;i<N;i++){
        scanf("%s %d",names[i],&expDate[i]);
    }
    bubbleSort(names,expDate);

    hset->heaplist[hset->size]=initheap(names[0]);
    input(hset->heaplist[hset->size],expDate[0]);
    hset->size++;
    for(int i=1;i<N;i++){
        if(strcmp(names[i],names[i-1])){
            hset->heaplist[hset->size]=initheap(names[i]);
            input(hset->heaplist[hset->size],expDate[i]);
            hset->size++;
        }
        else{
            input(hset->heaplist[hset->size-1],expDate[i]);
        }
    }
    printf("찾으시려는 품목은 \"G 품목명\"으로 입력, 프로그램을 종료하려면 \"Q\"입력\n");
    while(1){
        int foundName=0;
        scanf("%c",&cmd);
        if(cmd=='G'){
            scanf("%s",temp);
            for(int i=0;i<hset->size;i++){
                if(!strcmp(hset->heaplist[i]->name,temp)){
                    int deleted=delete(hset->heaplist[i]);
                    if(deleted==0)printf("-> 없어요!\n");
                    else{
                        printf("-> %s %d\n",hset->heaplist[i]->name,deleted);
                    }
                    foundName=1;
                    break;
                }
            }
            if(foundName==0)printf("-> 없어요!\n");
        }
        else if (cmd=='Q')break;
    }
    for(int i=0;i<N;i++){
        free(names[i]);
    }
    free(temp);
    free(names);
    free(expDate);
    return 0;
}
Heap* initheap(char* input){
    Heap* new=(Heap*)malloc(sizeof(Heap));
    new->size=0;
    strcpy(new->name,input);
    return new;
}
void swap(Heap* h, int a, int b){
    int temp;
    temp=h->heap[a].data;
    h->heap[a].data=h->heap[b].data;
    h->heap[b].data=temp;
}
void input(Heap* h, int num){
    h->size++;
    int i=h->size;
    h->heap[h->size].data=num;

    while(i>1){
        if(h->heap[i/2].data > h->heap[i].data){
            swap(h,i,i/2);
            i/=2;
        }
        else break;
    }
}
int delete(Heap* h){
    int value=h->heap[1].data;
    if(h->size==0){
        return 0;
    }
    h->heap[1].data=h->heap[h->size].data;
    if(h->size!=0)h->size--;
    int i=1,j;

    while(i<=h->size){
        j=i*2;
        if(j>h->size)break;
        if(h->heap[i].data < h->heap[j].data && h->heap[i].data < h->heap[j+1].data)break;
        if(h->heap[j].data > h->heap[j+1].data)j+=1;
        swap(h,i,j);
        i=j;
    }
    return value;
}
void bubbleSort(char** arr, int* arr2){
    char* temp;
    int tempNum;
    for(int i=0;i<N-1;i++){
        for(int j=0;j<N-1-i;j++){
            if(strcmp(arr[j],arr[j+1])>0){
                SWAP(arr[j],arr[j+1],temp);
                SWAP(arr2[j],arr2[j+1],tempNum);
            }
        }
    }
}

File: 트랜스폼컨쿼/applicationProblem.py
class min_heap:
    def __init__(self,name):
        self.heap=[None]
        self.name=name
    def input(self,data):
        self.heap.append(data)
        i=len(self.heap)-1
        while i>1:
            if self.heap[i]<self.heap[i//2]:
                self.heap[i],self.heap[i//2]=self.heap[i//2],self.heap[i]
                i=i//2
            else: break
    def delete(self):
        size=len(self.heap)-1
        if size==0: return 0
        value=self.heap[1]
        self.heap[1]=self.heap[size]
        self.heap.pop()
        if size!=0: size=size-1
        i=1
        while i<=size:
            j=i*2
            if j+1>size: break
            if self.heap[i] < self.heap[j] and self.heap[i] < self.heap[j+1]: break
            if self.heap[j] > self.heap[j+1]: j=j+1
            self.heap[i],self.heap[j]=self.heap[j],self.heap[i]
            i=j
        return value
    def print(self):
        for i in range(len(self.heap)):
            print(self.heap[i],end=' ')
class heapSET:
    def __init__(self):
        self.heaplist=[]
        self.size=0
    def newProduct(self,name):
        self.heaplist.append(min_heap(name))
        self.size+=1

def bubbleSort(namearr,datearr,n):
    for i in range(n-1):
        for j in range(n-1):
            if namearr[j] > namearr[j+1]:
                namearr[j],namearr[j+1]=namearr[j+1],namearr[j]
                datearr[j],datearr[j+1]=datearr[j+1],datearr[j]

names=[]
expDate=[]
hset=heapSET()
N=int(input())
for i in range(N):
    lst=input().split()
    names.append(lst[0])
    expDate.append(int(lst[1]))
bubbleSort(names,expDate,N)
hset.newProduct(names[0])
hset.heaplist[hset.size-1].input(expDate[0])
for i in range(1,N):
    if names[i]!=names[i-1]:
        hset.newProduct(names[i])
        hset.heaplist[hset.size-1].input(expDate[i])
    else:
        hset.heaplist[hset.size-1].input(expDate[i])
print("찾으시려는 품목은 \"G 품목명\"으로 입력, 프로그램을 종료하려면 \"Q\"입력")
while True:
    foundName=0
    cmd=input().split()
    if cmd[0]=='Q':
        break
    elif cmd[0]=='G':
        for i in range(hset.size):
            if cmd[1]==hset.heaplist[i].name:
                deleted=hset.heaplist[i].delete()
                if deleted==0: print("-> 없어요!")
                else: print("->",hset.heaplist[i].name,deleted)
                foundName=1
                break
        if foundName==0:
            print("-> 없어요!")
    


